Fix for:
16245956 M4: ROM Mapping Error for Matrox Device after First Run

Investigation shows during first run, option register's
biosen bit (the 30th bit) is reset to 0, which in turn
causes ROM base address to be reset to 0 (according to
matrox manual). The reason that we don't see this problem
on T5 systems and the sample matrox cards we were using
for development work, is the matrox chips on these devices
are of revision 4, while the matrox chips used on M4 systems
are of revision 5. Revision 4 uses SDRAM but revsion 5 uses
non-SDRAM (could be SGRAM). The initialization bits
for option register is different for them: revision 4
set biosen but revision 5 clears this bit.

The fix is to set pReg->Option whatever intended but keep the biosen
bit intact. This reduces impact it might have on other devices.

--- xf86-video-mga-2.0.0/src/mga_dacG.c
+++ xf86-video-mga-2.0.0/src/mga_dacG.c
@@ -1264,6 +1264,19 @@ MGA1064_MISC_CTL_DAC_RAM_CS;
 
 		if (pMga->HasSDRAM)
 		    pReg->Option = 0x40049120;
+		else {
+		    /* 
+		     * Do not reset biosen bit of option reg, as this will set 
+		     * ROM base register to 0
+		     */
+		    CARD32 option = 0;
+#ifdef XSERVER_LIBPCIACCESS
+		    pci_device_cfg_read_u32(pMga->PciInfo, & option, PCI_OPTION_REG);
+#else
+		    option = pciReadLong(pMga->PciTag, PCI_OPTION_REG);
+#endif
+		    pReg->Option = (pReg->Option & 0xFBFFFFFF) | (option & 0x40000000);
+		}
 	        pReg->Option2 = 0x00008000;
 		break;
 
